lib/hurd.c

```diff
@@ -0,0 +1,381 @@
+static void
+hurd_init_dev(struct pci_dev *d)
+{
+  d->aux = calloc(1, sizeof(mach_port_t));
+  assert(d->aux);
+}
+
+/* Deallocate the port and free its space */
+static void
+hurd_cleanup_dev(struct pci_dev *d)
+{
+  mach_port_t device_port;
+
+  device_port = *((mach_port_t *) d->aux);
+  mach_port_deallocate(mach_task_self(), device_port);
+
+  free(d->aux);
+}
+
+/* Walk through the FS tree to see what is allowed for us */
+static int
+enum_devices(const char *parent, struct pci_access *a, int domain, int bus,
+	     int dev, int func, tree_level lev)
+{
+  int err, ret;
+  DIR *dir;
+  struct dirent *entry;
+  char path[NAME_MAX];
+  char server[NAME_MAX];
+  uint32_t vd;
+  uint8_t ht;
+  mach_port_t device_port;
+  struct pci_dev *d;
+
+  dir = opendir(parent);
+  if (!dir)
+    return errno;
+
+  while ((entry = readdir(dir)) != 0)
+    {
+      snprintf(path, NAME_MAX, "%s/%s", parent, entry->d_name);
+      if (entry->d_type == DT_DIR)
+	{
+	  if (!strncmp(entry->d_name, ".", NAME_MAX)
+	      || !strncmp(entry->d_name, "..", NAME_MAX))
+	    continue;
+
+	  errno = 0;
+	  ret = strtol(entry->d_name, 0, 16);
+	  if (errno)
+	    {
+	      closedir(dir);
+	      return errno;
+	    }
+
+	  /*
+	   * We found a valid directory.
+	   * Update the address and switch to the next level.
+	   */
+	  switch (lev)
+	    {
+	    case LEVEL_DOMAIN:
+	      domain = ret;
+	      break;
+	    case LEVEL_BUS:
+	      bus = ret;
+	      break;
+	    case LEVEL_DEV:
+	      dev = ret;
+	      break;
+	    case LEVEL_FUNC:
+	      func = ret;
+	      break;
+	    default:
+	      if (closedir(dir) < 0)
+		return errno;
+	      return -1;
+	    }
+
+	  err = enum_devices(path, a, domain, bus, dev, func, lev + 1);
+	  if (err && err != EPERM && err != EACCES)
+	    {
+	      if (closedir(dir) < 0)
+		return errno;
+	      return err;
+	    }
+	}
+      else
+	{
+	  if (strncmp(entry->d_name, FILE_CONFIG_NAME, NAME_MAX))
+	    /* We are looking for the config file */
+	    continue;
+
+	  /* We found an available virtual device, add it to our list */
+	  snprintf(server, NAME_MAX, "%s/%04x/%02x/%02x/%01u/%s",
+		   _SERVERS_BUS_PCI, domain, bus, dev, func, entry->d_name);
+	  device_port = file_name_lookup(server, 0, 0);
+	  if (device_port == MACH_PORT_NULL)
+	    {
+	      closedir(dir);
+	      return errno;
+	    }
+
+	  d = pci_alloc_dev(a);
+	  *((mach_port_t *) d->aux) = device_port;
+	  d->bus = bus;
+	  d->dev = dev;
+	  d->func = func;
+	  pci_link_dev(a, d);
+
+	  vd = pci_read_long(d, PCI_VENDOR_ID);
+	  ht = pci_read_byte(d, PCI_HEADER_TYPE);
+
+	  d->vendor_id = vd & 0xffff;
+	  d->device_id = vd >> 16U;
+	  d->known_fields = PCI_FILL_IDENT;
+	  d->hdrtype = ht;
+	}
+    }
+
+  if (closedir(dir) < 0)
+    return errno;
+
+  return 0;
+}
+
+/* Enumerate devices */
+static void
+hurd_scan(struct pci_access *a)
+{
+  int err;
+
+  err = enum_devices(_SERVERS_BUS_PCI, a, -1, -1, -1, -1, LEVEL_DOMAIN);
+  assert(err == 0);
```

Comment:
Please do not use asserts. Report errors via the a->error callback, which can point to a custom error handler provided by the user of the library.

Reply:
Done in d1f0616